util/unifyfs/src/unifyfs-rm.c

```diff
@@ -244,6 +256,111 @@ static int wait_server_initialization(unifyfs_resource_t* resource,
     return ret;
 }
 
+enum {
+    UNIFYFS_STAGE_IN = 0,
+    UNIFYFS_STAGE_OUT = 1,
+};
+
+static inline unsigned int estimate_timeout(const char* manifest_file, int op)
+{
+    /* FIXME: we just wait for 20 mins.
+     * In fact, we can roughly estimate time for stage in by looking at the
+     * total file size. However, such a method is not possible for stage out
+     * because we have no idea on what the total transfer size is.
+     */
+    return 20 * 60;
+}
+
+
+/**
+ * @brief wait until data stage operation finishes
+ *
+ * @param resource
+ * @param args
+ *
+ * @return
+ */
+static
+int wait_stage(unifyfs_resource_t* resource, unifyfs_args_t* args, int op)
+{
+    int ret = UNIFYFS_SUCCESS;
+    unsigned int interval = 5;
+    unsigned int wait_time = 0;
+    unsigned int timeout = 0;
+    FILE* fp = NULL;
+    const char* manifest_file = NULL;
+    char filename[PATH_MAX];
+    char linebuf[16];
+    int return_val_from_scnprintf;
+
+    return_val_from_scnprintf =
+        scnprintf(filename, PATH_MAX,
+                  "%s/%s", args->share_dir, UNIFYFS_STAGE_STATUS_FILENAME);
+    if (return_val_from_scnprintf > (PATH_MAX - 2)) {
+        fprintf(stderr, "Unifyfs status filename is too long!\n");
+        return -ENOMEM;
+    }
+
+
+    if (op == UNIFYFS_STAGE_IN) {
+        manifest_file = args->stage_in;
+    } else {
+        manifest_file = args->stage_out;
+    }
+
+    if (args->stage_timeout > 0) {
+        timeout = args->stage_timeout;
+    } else {
+        timeout = estimate_timeout(manifest_file, op);
+    }
+
+    while (1) {
+        fp = fopen(filename, "r");
+        if (fp) {
+            char* line = fgets(linebuf, 15, fp);
+            if (0 == strncmp("success", line, strlen("success"))) {
+                ret = 0;
+                break;      // transfer completed
+            } else if (0 == strncmp("fail", line, strlen("fail"))) {
+                ret = -EIO;
+                break;      // transfer failed
+            } else {
+                fclose(fp); // try again
+            }
+        }
+
+
+        if (errno != ENOENT) {
+            fprintf(stderr, "failed to open file %s (%s)\n",
+                    UNIFYFS_STAGE_STATUS_FILENAME, strerror(errno));
+            ret = -errno;
+            break;
+        }
+
+        wait_time += interval;
+        sleep(interval);
+
+        if (wait_time > timeout) {
+            ret = UNIFYFS_FAILURE;
+            break;
+        }
+    }
+
+    return ret;
+}
+
+static inline int wait_stage_in(unifyfs_resource_t* resource,
+                                unifyfs_args_t* args)
+{
+    return wait_stage(resource, args, UNIFYFS_STAGE_IN);
+}
+
+static inline int wait_stage_out(unifyfs_resource_t* resource,
```

Comment:
These should not be inlined

Reply:
Would you mind explaining why they should not?

Reply:
Tony?  Any thoughts on this?  Why shouldn't these functions be inlined?  
I don't care either way, but it seems to be a reasonable use of inlining.  It lowers the duplication of code by having the lower function called with two different function wrappers but without the added context switch induced by calling one function with another one.

Reply:
https://www.kernel.org/doc/html/latest/process/coding-style.html#the-inline-disease
In this case the function is static, so the compiler will already optimize it.

Reply:
@tonyhutter I see your point there, but isn't it more like it "doesn't have to" be inlined? Like Craig mentioned, it is not critical whether these functions become inline or not, but I don't think it makes much sense to say "inline keyword is there to give a hint to the compiler to inline the function when possible, but you shouldn't be using it because the compiler is going to do it anyway when possible." The document you linked basically says "use it appropriately."  These functions have a single line, which one could go with macros instead. So, I don't see any problems with specifying the inline keyword there. Also, the compiler would not inline a function when not appropriate even with the inline keyword.

Reply:
Ok, I think that's fair then if you want to inline it to signify "this is a function that's basically a macro".